File: packages/sdk-go/openai_adapter.go
package agentshield

import (
	"encoding/json"
)

type OpenAIToolCall struct {
	ID       string `json:"id"`
	Type     string `json:"type"`
	Function struct {
		Name      string `json:"name"`
		Arguments string `json:"arguments"`
	} `json:"function"`
}

type OpenAIAdapter struct {
	shield *AgentShield
}

func NewOpenAIAdapter(shield *AgentShield) *OpenAIAdapter {
	if shield == nil {
		shield = New(AgentShieldConfig{})
	}
	return &OpenAIAdapter{shield: shield}
}

type ValidationResult struct {
	ValidToolCalls    []OpenAIToolCall `json:"validToolCalls"`
	BlockedToolCalls  []BlockedToolCall `json:"blockedToolCalls"`
}

type BlockedToolCall struct {
	ToolCall OpenAIToolCall `json:"toolCall"`
	Reason   string        `json:"reason"`
}

func (a *OpenAIAdapter) ValidateToolCalls(toolCalls []OpenAIToolCall, policies map[string]GuardrailPolicy) ValidationResult {
	var valid []OpenAIToolCall
	var blocked []BlockedToolCall

	for _, tc := range toolCalls {
		toolName := tc.Function.Name
		policy, exists := policies[toolName]
		if !exists {
			policy = policies["*"]
		}

		var parsedParams map[string]interface{}
		if tc.Function.Arguments != "" {
			json.Unmarshal([]byte(tc.Function.Arguments), &parsedParams)
		}
		if parsedParams == nil {
			parsedParams = map[string]interface{}{"raw": tc.Function.Arguments}
		}

		evalResult := a.shield.Guard(ToolCallRequest{
			ToolName: toolName,
			Params:   parsedParams,
		}, policy)

		if evalResult.Allowed {
			valid = append(valid, tc)
		} else {
			blocked = append(blocked, BlockedToolCall{
				ToolCall: tc,
				Reason:   evalResult.Reason,
			})
		}
	}

	return ValidationResult{
		ValidToolCalls:   valid,
		BlockedToolCalls: blocked,
	}
}

File: packages/sdk-rust/src/openai_adapter.rs
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenAIToolCall {
    pub id: String,
    #[serde(rename = "type")]
    pub type_field: String,
    pub function: OpenAIFunction,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenAIFunction {
    pub name: String,
    pub arguments: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ValidationResult {
    pub valid_tool_calls: Vec<OpenAIToolCall>,
    pub blocked_tool_calls: Vec<BlockedToolCall>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BlockedToolCall {
    pub tool_call: OpenAIToolCall,
    pub reason: String,
}

pub struct OpenAIAdapter {
    shield: crate::agentshield::AgentShield,
}

impl OpenAIAdapter {
    pub fn new(shield: crate::agentshield::AgentShield) -> Self {
        Self { shield }
    }

    pub fn validate_tool_calls(&self, tool_calls: Vec<OpenAIToolCall>, policies: HashMap<String, crate::types::GuardrailPolicy>) -> ValidationResult {
        let mut valid = Vec::new();
        let mut blocked = Vec::new();

        for tool_call in tool_calls {
            let tool_name = tool_call.function.name.clone();
            let policy = policies.get(&tool_name)
                .cloned()
                .or_else(|| policies.get("*").cloned())
                .unwrap_or_default();

            let mut parsed_params = HashMap::new();
            if !tool_call.function.arguments.is_empty() {
                if let Ok(v) = serde_json::from_str::<Value>(&tool_call.function.arguments) {
                    if let Value::Object(map) = v {
                        parsed_params = map.into_iter().collect();
                    }
                } else {
                    parsed_params.insert("raw".to_string(), Value::String(tool_call.function.arguments.clone()));
                }
            }

            let request = crate::types::ToolCallRequest {
                tool_name: tool_name.clone(),
                params: parsed_params,
                ..Default::default()
            };

            let eval_result = self.shield.guard(request, policy);

            if eval_result.allowed {
                valid.push(tool_call);
            } else {
                blocked.push(BlockedToolCall {
                    tool_call,
                    reason: eval_result.reason.unwrap_or_else(|| "Blocked by policy".to_string()),
                });
            }
        }

        ValidationResult {
            valid_tool_calls: valid,
            blocked_tool_calls: blocked,
        }
    }
}